src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +673,482 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
     }
 };
 
+class fetch_worker {
+public:
+    // Passed from the coordinator shard to fetch workers.
+    // Contains either references to objects local to the fetch worker shard or
+    // copies of data from the coordinator shard.
+    struct shard_local_fetch_context {
+        // True if the results of this fetch will be read on a foreign shard.
+        bool foreign_read;
+        size_t bytes_left;
+        // Specifies the minimum number of bytes this sub-fetch should read
+        // before returning.
+        size_t min_bytes;
+        // If set then the sub-fetch should return by the specified time_point .
+        std::optional<model::timeout_clock::time_point> deadline;
+        // The fetch sub-requests of partitions local to the shard this worker
+        // is running on.
+        std::vector<ntp_fetch_config> requests;
+
+        // References to services local to the shard this worker is running on.
+        // They are protected from deletion by the coordinator.
+        server& srv;
+        cluster::partition_manager& mgr;
+        ss::abort_source& as;
```

Comment:
I don't see what value we are getting from the use of sharded abort source for the specific purpose of communicating from the coordinator to the worker that they should stop. 

sharded about source is basically two parts: the ss::abort_source machinery, plus the ss::sharded_serivce machinery. 

The sharded_service itself is just a list of shard-local services in a vector and some stuff to start and stop them. I don't thing we get any value from it: the vector of services is the wrong size: we want 1 per worker shard, not 1 per shard on the broker and we already have our list of coordinator shards explicitly in the coordinator. Furthermore, the start/stop stuff is not useful as it's just a simple object we are talking about.

Then the ss::abort_source machinery is necessary on the worker since that's what hwm monitor uses, but I don't see why we need it to communicate the stop signal across shards: it's not like we can use this cross-shard abort source directly: we have to chain it anyway to the local one. Then there's the issue that the lifetime semantics of sharded_abort_source are wrong for us (and as you point out probably in other cases as well). So it seems like it's more trouble than it's worth.

Why don't we just directly sent a message to the right shards using invoke_on? When we create the shard fetch wokers, we can create the abort source on the coordinator side, share it with the worker, and when we need to stop, trigger the abort source on the worker using invoke_on. This abort source can replace the local abort source on the worker as well.

Reply:
Good point, switched to a vector of abort sources on the coordinator that are shared with the workers. This should allow us to reduce cross-shard calls as you've said.